source/opt/types.cpp

```diff
@@ -328,25 +328,43 @@ Image::Image(Type* type, SpvDim dimen, uint32_t d, bool array, bool multisample,
       ms_(multisample),
       sampled_(sampling),
       format_(f),
-      access_qualifier_(qualifier) {
+      access_qualifier_(qualifier),
+      has_access_qualifier_(true) {
   // TODO(antiagainst): check sampled_type
 }
 
+Image::Image(Type* type, SpvDim dimen, uint32_t d, bool array, bool multisample,
+             uint32_t sampling, SpvImageFormat f)
+    : Type(kImage),
+      sampled_type_(type),
+      dim_(dimen),
+      depth_(d),
+      arrayed_(array),
+      ms_(multisample),
+      sampled_(sampling),
+      format_(f),
+      access_qualifier_(SpvAccessQualifierReadOnly),
+      has_access_qualifier_(false) {}
+
 bool Image::IsSameImpl(const Type* that, IsSameCache* seen) const {
   const Image* it = that->AsImage();
   if (!it) return false;
   return dim_ == it->dim_ && depth_ == it->depth_ && arrayed_ == it->arrayed_ &&
          ms_ == it->ms_ && sampled_ == it->sampled_ && format_ == it->format_ &&
          access_qualifier_ == it->access_qualifier_ &&
```

Comment:
In other parts, you ignore `access_qualifier_` if `has_access_qualifer_` is false.  Why not here as well?